pkg/runtime/serializer/protobuf/protobuf.go

```diff
@@ -0,0 +1,418 @@
+// +build proto
+
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package protobuf
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"reflect"
+	"sync"
+
+	"github.com/gogo/protobuf/proto"
+
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/runtime"
+)
+
+var (
+	// protoEncodingPrefix serves as a magic number for an encoded protobuf message on this serializer. All
+	// proto messages serialized by this schema will be preceeded by the bytes 0x6b 0x38 0x73, with the fourth
+	// byte being reserved for the encoding style. The only encoding style defined is 0x00, which means that
+	// the rest of the byte stream is a message of type k8s.io.kubernetes.pkg.runtime.Unknown (proto2).
+	//
+	// See k8s.io/kubernetes/pkg/runtime/generated.proto for details of the runtime.Unknown message.
+	//
+	// This encoding scheme is experimental, and is subject to change at any time.
+	protoEncodingPrefix = []byte{0x6b, 0x38, 0x73, 0x00}
```

Comment:
We can't define slices as const unfortunately.  We could do it as a string
but that's kind of ugly.

On Mon, Mar 7, 2016 at 7:36 AM, Wojciech Tyczynski <notifications@github.com
> wrote:

> In pkg/runtime/serializer/protobuf/protobuf.go
> <https://github.com/kubernetes/kubernetes/pull/20377#discussion_r55203238>
> :
>
> > +	"github.com/gogo/protobuf/proto"
> > +
> > +	"k8s.io/kubernetes/pkg/api/unversioned"
> > +	"k8s.io/kubernetes/pkg/runtime"
> > +)
> > +
> > +var (
> > +	// protoEncodingPrefix serves as a magic number for an encoded protobuf message on this serializer. All
> > +	// proto messages serialized by this schema will be preceeded by the bytes 0x6b 0x38 0x73, with the fourth
> > +	// byte being reserved for the encoding style. The only encoding style defined is 0x00, which means that
> > +	// the rest of the byte stream is a message of type k8s.io.kubernetes.pkg.runtime.Unknown (proto2).
> > +	//
> > +	// See k8s.io/kubernetes/pkg/runtime/generated.proto for details of the runtime.Unknown message.
> > +	//
> > +	// This encoding scheme is experimental, and is subject to change at any time.
> > +	protoEncodingPrefix = []byte{0x6b, 0x38, 0x73, 0x00}
>
> nit: should we define it as const?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20377/files#r55203238>.
>